### ISSUE TITLE
2.0 Reference Models: BERT (TPU with dist strat and Keras)

### ISSUE THREAD

**Issue by dynamicwebpaige** [tensorflow/tensorflow#26994]
[**BERT**, or Bidirectional Encoder Representations from Transformers](https://github.com/google-research/bert), is a new method of pre-training language representations which obtains state-of-the-art results on a wide array of Natural Language Processing (NLP) tasks.

The research team's academic paper describes BERT in detail and provides full results on a number of tasks: https://arxiv.org/abs/1810.04805.

An example of using BERT can be found [here](https://colab.sandbox.google.com/github/google-research/bert/blob/master/predicting_movie_reviews_with_bert_on_tf_hub.ipynb).

The purpose of this issue is to migrate models into the TF 2.0 Keras API. Each migrated model must be eager and distribution compatible, with tests, and all associated engineering artifacts.

**Comment by rok** [comment]
@dynamicwebpaige I'd like to contribute this.

**Comment by icanswim** [comment]
@dynamicwebpaige 
very interested in this.  how can help?

**Comment by icanswim** [comment]
@rok 
are you working on this?  i think ill start.  would you like to collaborate?  im just getting into TF2..

**Comment by rok** [comment]
Hey @icanswim! Sure, it would be a pleasure.
I didn't properly start yet because TPU support for TF2 seems to be lacking right now.
I gave you access to [my fork](https://github.com/rok/models) or we can work via yours. I'll ping when I have some work done.

**Comment by rok** [comment]
Hey @icanswim, I've done some more work on top of your commit and I'll do some more (tests are passing but there are still warnings).

I already [opened a PR](https://github.com/tensorflow/models/pull/6817), could you please sign the CLA there so it can be reviewed?

**Comment by tfboyd** [comment]
Sorry we missed this.  We are working on this internally and are almost done.  Expect code to be pushed before end of May-2019.

**Comment by ericpts** [comment]
Are there any updates on this?

**Comment by tfboyd** [comment]
https://github.com/tensorflow/models/tree/master/official/bert

The README has not been updated yet, but I think the information below will get you started.

I am only testing the GPU version but it was build for TPU so that would have been tested first.  I do not have the flags to use but I have the test cases we run nightly.  Below is what I know, I have not been going even a little deep into BERT so I don't know how to state the accuracy numbers other than what we are tracking nightly.

BERT (SQUAD):  .0398 to .0439
BERT (Classify):  .8603 with mrpc

   * [Test Case SQUAD](https://github.com/tensorflow/models/blob/master/official/bert/benchmark/bert_squad_benchmark.py): official.bert.benchmark.bert_squad_benchmark.BertSquadBenchmarkReal.benchmark_8_gpu
   * [Test Case Classify](https://github.com/tensorflow/models/blob/master/official/bert/benchmark/bert_benchmark.py): official.bert.benchmark.bert_benchmark.BertClassifyAccuracy.benchmark_8_gpu_mrpc

I realize it is a bit painful but you can extract the FLAGS from the tests above. We are still working on performance but we are already exceeding the TF 1.0 for both single and multi-gpu single machine.  FP16 support will be coming soon.  We are working on SQUAD first but the gains would apply to both...in most instances.  

**Warning:** The data in gs://cloud-tpu-checkpoints is public and you can pull that down.  I did not make the actual training data public because I have not taken time to find out if it can be shared directly.  It likely can, but I like to be careful.

I will help if I can; another group owns this but I deal with it daily as well for now.

**Comment by saberkun** [comment]
We will publish tutorial/colab to help go through the training process.
I will first update README to explain basics like checkpoints and training data generation. (ETA: by the end of this week)

**Comment by dynamicwebpaige** [comment]
Thank you, @saberkun! Closing out this issue - and excellent work! 😄
